### ISSUE TITLE
cmd/compile: internal compiler error: dictionary variable not captured by value

### ISSUE THREAD

**Issue by believening** [golang/go#60569]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.19.9 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

worked in 1.20

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/xxx/.cache/go-build"
GOENV="/home/xxx/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/xxx/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/xxx/go"
GOPRIVATE=""
GOPROXY=""
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19.9"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1259709727=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

A simplified version of our actual code compiles successfully in 1.20, but not in 1.19.

https://go.dev/play/p/-rFoI3jndSq?v=goprev

### What did you expect to see?

build successfully

### What did you see instead?

build failed

```
internal compiler error: dictionary variable not captured by value

Please file a bug report including a short program that triggers the error.
https://go.dev/issue/new
```

**Comment by mknyszek** [comment]
CC @mdempsky @griesemer

**Comment by mknyszek** [comment]
CC @golang/compiler

**Comment by mdempsky** [comment]
Go 1.20 uses a completely different implementation of generics than Go 1.19 did. Backporting a fix for this to Go 1.19 is impractical, sorry.
